source/extensions/filters/network/kafka/kafka_request_parser.h

```diff
@@ -0,0 +1,197 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/common/exception.h"
+
+#include "common/common/assert.h"
+
+#include "extensions/filters/network/kafka/kafka_request.h"
+#include "extensions/filters/network/kafka/parser.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace Kafka {
+
+/**
+ * Context that is shared between parsers that are handling the same single message
+ */
+struct RequestContext {
+  int32_t remaining_request_size_{0};
+  RequestHeader request_header_{};
+};
+
+typedef std::shared_ptr<RequestContext> RequestContextSharedPtr;
+
+/**
+ * Configuration object
+ * Resolves the parser that will be responsible for consuming the request-specific data
+ * In other words: provides (api_key, api_version) -> Parser function
+ */
+class RequestParserResolver {
+public:
+  virtual ~RequestParserResolver() = default;
+
+  /**
+   * Creates a parser that is going to process data specific for given api_key & api_version
+   * @param api_key request type
+   * @param api_version request version
+   * @param context context to be used by parser
+   * @return parser that is capable of processing data for given request type & version
+   */
+  virtual ParserSharedPtr createParser(int16_t api_key, int16_t api_version,
+                                       RequestContextSharedPtr context) const;
+
+  /**
+   * Request parser singleton
+   */
+  static const RequestParserResolver INSTANCE;
```

Comment:
nit: s/INSTANCE/instance, but better, make this a function and use CONSTRUCT_ON_FIRST_USE.